src/StateMachine/StateMachine.ts

```diff
@@ -250,6 +250,20 @@ export class StateMachine {
       throw new UnknownState(currentStateName);
     }
 
-    return states[0];
+    if (states.length === 1) {
+      return states[0];
+    }
+
+    // If the code reaches this points that means the `states` array may contains
+    // one state without an intents array filter and/or
+    // one or more states with an intents array that contains the intent name.
+    // The state with an intents array is more important than the one with no intents array,
+    // so let's return the first state that contains the intent name in its intents array.
+
+    return (
+      states.find(
```

Comment:
i'm actually thinking we should rename this to "StateController" or something like that :joy: